addons/uom/models/uom_uom.py

```diff
@@ -4,20 +4,70 @@
 from odoo import api, fields, tools, models, _
 from odoo.exceptions import UserError, ValidationError
 
+def _check_category_reference_uniqueness(cat):
+    cnt=sum(uom.uom_type=='reference' and uom.active for uom in cat.uom_ids)
+    if cnt > 1:
+        raise ValidationError(_("UoM category %s should only have one reference unit of measure.") % cat.name)
+    elif cnt==0:
+        raise ValidationError(_("UoM category %s should have a reference unit of measure.") % cat.name)
 
 class UoMCategory(models.Model):
     _name = 'uom.category'
     _description = 'Product UoM Categories'
 
     name = fields.Char('Unit of Measure Category', required=True, translate=True)
 
+    uom_ids = fields.One2many('uom.uom', 'category_id')
+    locked = fields.Boolean("Locked", store=False)
+
+    @api.model
+    def create(self, values):
+        rec = super(UoMCategory, self.with_context(no_unique_reference_check=True)).create(values)
+        if len(rec.uom_ids)>1:
+            _check_category_reference_uniqueness(rec)
+        return rec
+ 
+    def write(self, values):
+        super(UoMCategory, self.with_context(no_unique_reference_check=True)).write(values)
+        if len(self.uom_ids)>1:
+            _check_category_reference_uniqueness(self)
+
     def unlink(self):
         uom_categ_unit = self.env.ref('uom.product_uom_categ_unit')
         uom_categ_wtime = self.env.ref('uom.uom_categ_wtime')
         if any(categ.id in (uom_categ_unit + uom_categ_wtime).ids for categ in self):
             raise UserError(_("You cannot delete this UoM Category as it is used by the system."))
         return super(UoMCategory, self).unlink()
 
+    @api.onchange('uom_ids')
+    def _onchange_uom_ids(self):
+        if len(self.uom_ids) == 1:
+            self.uom_ids[0].uom_type = 'reference'
+            self.uom_ids[0].factor = 1
+        elif len(self.uom_ids) > 1:
+            cnt=sum(uom.uom_type=='reference' and uom.active for uom in self.uom_ids)
+            if cnt==0:
+                warning = {
+                    'title': _('Warning!'),
+                    'message': _("UoM category %s should have a reference unit of measure.") % self.name
+                }
+                return {'warning': warning}
+            if cnt > 1:
+                newref = self.uom_ids.filtered(lambda o : o.uom_type == 'reference' and o.active and o._origin.uom_type != 'reference')
+                if newref:
+                    others = self.uom_ids - newref
+                    for o in others:
+                        v = {}
+                        v['factor'] = o.factor / (newref._origin.factor or 1)
+                        if v['factor'] > 1:
+                            v['uom_type'] = 'smaller'
+                        else:
+                            v['uom_type'] = 'bigger'
+                        o.update({'uom_type': v['uom_type']})
```

Comment:
uom.update({
    'key1': 'value1',
    'key2': 'value2',
    ...
})
Since you are in an onchange you can directly assign to UoM, it will be an implicit update.
uom.uom_type = 'smaller'